### PR TITLE
[HZ-333] Support IMap SQL INSERTs

### PR DESCRIPTION
**Pull request by gierlachg**
Added support for `IMap` `INSERT`s.

`INSERT` contrary to `SINK` fails on duplicate key, whether entry with given key already existed or is inserted twice. To track _current `INSERT`_ keys, they are accumulated in the processor. Hence, `IMap` `INSERT` is subject to `maxProcessorAccumulatedRecords` limit.

### CODE REVIEW

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertProcessorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeString(mapName);
+        out.writeObject(projectorSupplier);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        mapName = in.readString();
+        projectorSupplier = in.readObject();
+    }
```
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertProcessorSupplier.java:84]: If I understand correctly that would require using `ProcessorSupplier.of()`. However, we don't support `IdentifiedDataSerializable` inside lambdas.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java [hunk 2]
```diff
@@ -283,24 +288,19 @@ default VertexWithInputConfig nestedLoopReader(
...
      * Returns the supplier for the sink processor.
      */
     @Nonnull
-    default Vertex sinkProcessor(
-            @Nonnull DAG dag,
-            @Nonnull Table table
-    ) {
-        throw new UnsupportedOperationException("INSERT INTO or SINK INTO not supported for " + typeName());
+    default Vertex sinkProcessor(@Nonnull DAG dag, @Nonnull Table table) {
+        throw new UnsupportedOperationException("SINK INTO not supported for " + typeName());
```
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java:298]: What do you propose exactly? We need to configure the edge inside `IMap`s `insertProcessor()`.

[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java:298]: ok, my mistake.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertProcessorSupplier.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertProcessorSupplier.java]: https://github.com/hazelcast/hazelcast/pull/18940/commits/975b01bd2fe9bec3cb12897605c3e83f9c53c482

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java:298]: ok, my mistake

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertP.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        projector = projectorSupplier.get(serializationService);
+    }
+
+    @Override
+    protected boolean tryProcess(int ordinal, @Nonnull Object row) {
+        if (isQueueFull()) {
+            if (!tryFlushQueue() && isQueueFull()) {
+                return false;
+            }
+        }
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/InsertP.java:69]: What about this?
```suggestion
        tryFlushQueue();
        if (isQueueFull()) {
            return false;
        }
```
We'll make room for more requests even if the queue isn't fully flushed.

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**